### ISSUE TITLE
timer

### ISSUE THREAD

**Issue by davidszotten** [onefinestay/nameko#303]
The `Timer` claims to 

```
# next time, sleep however long is left of our interval, taking
# off the time we took to run
```

but we only measure the time of `spawn_worker`, not the actual entrypoint run

looks like this has been broken a very long time

try e.g.

```
import time
from nameko.timer import timer

clock = time.time()


class Service(object):
    name = 'timertest'

    @timer(1)
    def tick(self):
        print 'tick', time.time() - clock
        time.sleep(5)
```

we should either remove the broken "feature", or if we want this despite living without it for this long, fix it (might be a bit of a faff)

**Comment by Sparkycz** [comment]
I'm afraid i don't understand you.. 
You want to wait on finish of last round and then new one will start, don't you?

**Comment by davidszotten** [comment]
the problem is that in the current implementation, we measure how long `self.handle_timer_tick()`. but that function calls `spawn_worker` which is asynchronous, and returns as soon as the handler is _spawned_, and doesn't wait for it to finish.

in the example above, you see `'tick'` every second, even though the handler takes 5 seconds to run

**Comment by mattbennett** [comment]
I'm in favour of fixing this. I think the semantics should be either:

* At most once per `interval` seconds or
* Exactly `interval` seconds after the previous execution finishes

Looks like @Sparkycz's PR does the second option.

**Comment by davidszotten** [comment]
to me it looks like the pr waits `interval - elapsed_time`, which is what the original code tried to do, and my preferred choice i think. is this your option 1?

**Comment by Sparkycz** [comment]
the both PRs are wrong.. i'll send the first one now and meanwhile i'm trying to solve the second one. I'm sorry I havn't have the time for it..

**Comment by Sparkycz** [comment]
Currently it probably does what we want. If the service method works longer than interval Timer waits for the completion and then fires the next one.

**Comment by mattbennett** [comment]
@davidszotten Yep, waiting `interval - elapsed_time` is Option 1 and also my preference.

@Sparkycz I will add a few comments to the PR. Please address them and then we'll merge this down. Thanks!

**Comment by daviskirk** [comment]
Is this still being worked on?  If it's just a matter of finishing up the tests I can give it a try

**Comment by mattbennett** [comment]
@daviskirk I think almost certainly not. A fix would be great!

**Comment by daviskirk** [comment]
@mattbennett see #579
